csrc/val_graph_visitor.cpp

```diff
@@ -0,0 +1,150 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <val_graph_visitor.h>
+
+#include <id_model/to_string.h>
+
+namespace nvfuser {
+
+void ValGraphVisitor::traverse() {
+  const ValGroups terminating_inputs = graph().getTerminatingInputs();
+  std::deque<ValGroup> to_visit_vals(
+      terminating_inputs.begin(), terminating_inputs.end());
+  ValGroups visited_vals;
+
+  std::deque<ExprGroup> to_visit_exprs;
+  ExprGroups visited_exprs;
+
+  auto is_expr_ready = [&](const ExprGroup& expr_group) -> bool {
+    const auto inp_groups = graph().inputGroups(expr_group);
+    return std::all_of(
+        inp_groups.begin(), inp_groups.end(), [&](ValGroup val_group) {
+          return visited_vals.has(val_group) || val_group->empty();
+        });
+  };
+
+  // If any input of the def expr is mapped with the val
+  // group itself, i.e., a trivial expr, allow visiting the
+  // val group first. The trivial expr group will be visited
+  // after the val group.
+  //
+  // Example:
+  //
+  // [i0, 1]
+  // merge
+  // [i0*1]
+  // map i0 and i0*1
+  // ValGroups: {{i0, i0*1}, {1}}
+  //
+  // Then, {i0, i0*1} and {1} would be visited first, then the merge
+  // expr group would be visited. {i0, i0*1} is also an output group
+  // of the merge but since it's already in the visited set, it would
+  // not be visited again.
+  //
+  // See also IdModelTest.ValGraphStmtSort3 for a concrete example.
+  auto is_val_ready = [&](const ValGroup& val_group) -> bool {
+    const ExprGroups& unique_defs = graph().getDefinitions(val_group);
+    return std::all_of(
+        unique_defs.begin(), unique_defs.end(), [&](ExprGroup expr_group) {
+          if (expr_group->empty() || visited_exprs.has(expr_group) ||
+              terminating_inputs.has(val_group)) {
+            return true;
+          }
+          // Handle trivial expr groups. This expr_group is not
+          // visited yet, which means there're input ValGroups that
+          // are not yet visited. If those not-visited inputs are
+          // actually the same as val_group, visit val_group at this
+          // point to resolve the circular dependency.
```

Comment:
Exactly. At this point, the order doesn't matter as the traversal in the loop promotion analysis should not have these cycles. This may matter in the future when generating different mappings and graphs, but at this point we could also just make it fail if such cycles were detected. I'll add more notes.

Reply:
> at this point we could also just make it fail if such cycles were detected

That would be my preference. Instead of adding non-tested, arguable features, it's usually better to fail explicitly for now (if it simplifies).